### ISSUE TITLE
effectful generators are not unwound

### ISSUE THREAD

**Issue by symbiont-sam-halliday** [hedgehogqa/haskell-hedgehog#334]
As noted in https://github.com/hedgehogqa/haskell-hedgehog/pull/330 which was a purported fix for #287, when a test fails the effect is reused from that failure instead of the effect from the earlier branch in the search tree.

This means that if we use `StateT` to store some state, the same state that was seen in the failure case is used in the shrink. But I had expected that the state of the previous branch in the search tree would have been used, which might mean reverting back to the original state.

Consider the simple case where we want to build up a set of valid keys along with some data, the state contains `["foo", "bar", "baz"]` and then a test fails. The shrinker might remove the data generator that led us to add `baz` to this list, now we're generating data thinking that `baz` is a valid key but it is not... so our test may continue to fail because our data generators are now invalid.

Or even simpler, consider a counter that increments 1 every time a `Foo` is generated and we are generating `[Foo]`. If a test fails then the counter != length of the list.

Basically `StateT` is behaving the way one would expect `ReaderT (IORef Int) IO` to behave, not the way one expects `State` to behave.

I am investigating if `ReaderT` is also affected. If it is, I think that means the usecase described in https://youtu.be/AIv_9T0xKEo?t=1404 is no longer true.

**Comment by symbiont-sam-halliday** [comment]
Here is a minimisation of the problem... it only fails in the unexpected way sometimes

```module StateTest where

import Control.Monad.State
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import Prelude

type G a = StateT Int Gen a

data Foo = Foo Int deriving (Show, Eq)

gFoo :: G Foo
gFoo = do
  i <- get
  put $ 1 + i
  pure $ Foo i

gFooList :: G [Foo]
gFooList = Gen.list (Range.constant 1 10) gFoo

hprop_badstate :: Property
hprop_badstate =
  property $ do
    foos <- forAll $ evalStateT gFooList 0
    assert $ length foos < 5
```

Let's consider this case that failed for me:

```
            ✗ badstate failed at tests/StateTest.hs:26:5
              after 1 test and 4 shrinks.
            
                 ┏━━ tests/StateTest.hs ━━━
              22 ┃ hprop_badstate :: Property
              23 ┃ hprop_badstate =
              24 ┃   property $ do
              25 ┃     foos <- forAll $ evalStateT gFooList 0
                 ┃     │ [ Foo 4 , Foo 5 , Foo 6 , Foo 7 , Foo 8 ]
              26 ┃     assert $ length foos < 5
                 ┃     ^^^^^^^^^^^^^^^^^^^^^^^^
```

The `Foo`s are numbered 4 to 8. But to get the counter up to 3 in the first place, that means the state has been updated 3 times already before calling this generator. I would have expected the following to be the shrunken failure

```
                 ┃     │ [ Foo 0 , Foo 1 , Foo 2 , Foo 3 , Foo 4 ]

```

**Comment by symbiont-sam-halliday** [comment]
I'm unable to see the problem with a custom list generator (instead of `Gen.list`)

```
gList :: Range.Range Int -> G a -> G [a]
gList range gen = do
  n <- Gen.integral_ range
  replicateM n gen
```

but again this may be just avoiding the problem for this one example.

**Comment by symbiont-sam-halliday** [comment]
I suspect that the problem is to do with the triplet of helper functions in `Tree`

```
dropOne :: Monad m => [NodeT m a] -> [TreeT m [a]]
dropOne ts = do
  (xs, _y, zs) <- splits ts
  pure . TreeT . pure $
    interleave (xs ++ zs)

shrinkOne :: Monad m => [NodeT m a] -> [TreeT m [a]]
shrinkOne ts = do
  (xs, y0, zs) <- splits ts
  y1 <- nodeChildren y0
  pure . TreeT $ do
    y2 <- runTreeT y1
    pure $
      interleave (xs ++ [y2] ++ zs)

interleave :: forall m a. Monad m => [NodeT m a] -> NodeT m [a]
interleave ts =
  NodeT (fmap nodeValue ts) $
    concat [
        dropOne ts
      , shrinkOne ts
      ]
```

The call to `runTreeT` in `shrinkOne` is particularly suspicious because this `y2` can be removed by a recursion back into `shrinkOne` but the effect has still been evaluated. Hence why we see `State` changes.

**Comment by symbiont-sam-halliday** [comment]
I've spent all day trying to come up with an alternative to `Gen.list` that works for `State` but I have given up. I think it is just not possible to isolate this effect. Probably `NodeT` would need to be changed to have an `m a` for the value instead of `a` to stand a chance, but I'm not even sure if that's enough.

For now, I think the best solution is to document the behaviour: that generators of values depend on any effects that were performed while generating prior shrinks and the main value. That more or less reduces `State` to the same behaviour as one would get from `Reader (IORef a) IO`.

As a workaround, I'm trying to rewrite my `State` code to use `Reader` which will need a custom `Gen.list` taking an `(s, a) -> s` update function so that `local` can be applied to later values.

**Comment by symbiont-sam-halliday** [comment]
Here is a workaround using `ReaderT` instead of `StateT`, and generators need to be refactored so that updates to the state / content are passed when generating lists, in addition to other data structures:

```
gList :: Range Int -> G a -> (a -> Int -> Int) -> G [a]
gList range gen f = Gen.sized $ \size ->
  let lowest = Range.lowerBound size range
      -- We can only perform removal of elements from the end of the list. If we
      -- removed elements from the middle of the list then their impact on the
      -- reader context would still be felt.
      smaller = filter ((lowest <=) . length) . tailOrNil . reverse . inits
      tailOrNil = \case
        [] -> []
        _ : as -> as
  in Gen.shrink smaller $ do
    len <- Gen.integral_ range
    let rec 0 = pure []
        rec i = do
          a <- gen
          (a :) <$> local (f a) (rec $ i - 1)
    rec len
```

**Comment by symbiont-sam-halliday** [comment]
Closing due to lack of activity.
